pilot/pkg/xds/ecds.go

```diff
@@ -59,65 +59,92 @@ func ecdsNeedsPush(req *model.PushRequest) bool {
 	return false
 }
 
+// Check if the secret updates is relevant to Wasm image pull. If not relevant, skip pushing ECDS.
+// this occurs when user only update the Secret referenced by WasmPlugin/EnvoyFilter.
+func secretConfigsUpdatedOnly(req *model.PushRequest) bool {
+	return !model.ConfigsHaveKind(req.ConfigsUpdated, kind.WasmPlugin) &&
+		!model.ConfigsHaveKind(req.ConfigsUpdated, kind.EnvoyFilter) &&
+		model.ConfigsHaveKind(req.ConfigsUpdated, kind.Secret)
+}
+
 // Generate returns ECDS resources for a given proxy.
 func (e *EcdsGenerator) Generate(proxy *model.Proxy, w *model.WatchedResource, req *model.PushRequest) (model.Resources, model.XdsLogDetails, error) {
 	if !ecdsNeedsPush(req) {
 		return nil, model.DefaultXdsLogDetails, nil
 	}
 
+	secrets, needsPush := e.pullSecrets(proxy, w, req)
+	if !needsPush {
+		return nil, model.DefaultXdsLogDetails, nil
+	}
+
+	ec := e.Server.ConfigGenerator.BuildExtensionConfiguration(proxy, req.Push, w.ResourceNames, secrets)
+	if ec == nil {
+		return nil, model.DefaultXdsLogDetails, nil
+	}
+
+	resources := make(model.Resources, 0, len(ec))
+	for _, c := range ec {
+		resources = append(resources, &discovery.Resource{
+			Name:     c.Name,
+			Resource: protoconv.MessageToAny(c),
+		})
+	}
+
+	return resources, model.DefaultXdsLogDetails, nil
+}
+
+func (e *EcdsGenerator) pullSecrets(proxy *model.Proxy, w *model.WatchedResource, req *model.PushRequest) (secrets map[string][]byte, needsPush bool) {
 	secretResources := referencedSecrets(proxy, req.Push, w.ResourceNames)
-	// Check if the secret updates is relevant to Wasm image pull. If not relevant, skip pushing ECDS.
-	if !model.ConfigsHaveKind(req.ConfigsUpdated, kind.WasmPlugin) && !model.ConfigsHaveKind(req.ConfigsUpdated, kind.EnvoyFilter) &&
-		model.ConfigsHaveKind(req.ConfigsUpdated, kind.Secret) {
+	if secretConfigsUpdatedOnly(req) {
```

Comment:
IMO, this is an optimization to skip push ecds only when secret update, but the secret is not relevant. Add more comments

Reply:
Maybe better to move the if clause out